### ISSUE TITLE
Jetty 12: Implement HttpClient cache

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#11076]
**Jetty Version**
12.0.1

**Jetty Environment**
Client API

**Java Version**
21.0.1

**Question**
It's not clear to me how Jetty Client can respect `Cache-Control` headers without having a client-side cache to pull cached responses out of. But on the other hand, I can't find any sort of mention of a client-side cache in the documentation.

Can you please clarify what the client supports or does not support with respect to client-side caching?

**Comment by sbordet** [comment]
Jetty's `HttpClient` does not have a built-in cache.

**Comment by cowwoc** [comment]
@sbordet So, out of curiosity, how can it respect the `Cache-Control` header?

Does it integrate with external cache libraries? If so, where is this documented?

Thanks.

**Comment by sbordet** [comment]
> how can it respect the `Cache-Control` header?

I have not looked at this, but ideally you can record the responses via a listener, store them in a cache, and intercept requests being sent to not send them and return the cached response instead.
I think internally we have all the APIs to support that, but I'm not sure if they are public enough for third parties to implement a caching layer.

I am not aware of third party integrations with caching libraries.

In summary this would be a new feature, as it is currently not supported.

**Comment by cowwoc** [comment]
Okay, thanks. Can you please convert this issue into a feature request? Apache HttpClient seems to already have such a feature but I'd rather stick with an all-Jetty stack if possible.

**Comment by sbordet** [comment]
@cowwoc we can target Jetty 12, but when we got time.

If you need this feature urgently, consider [commercial support from Webtide](https://webtide.com).

**Comment by cowwoc** [comment]
It's not urgent and I'm currently unemployed, so I guess I'll have to wait :)

**Comment by sbordet** [comment]
> It's not urgent and I'm currently unemployed, so I guess I'll have to wait :)

How about contributing it? 😉

**Comment by cowwoc** [comment]
@sbordet I'll happily circle back once I'm done my current project, but it might take a while. I'm trying to start my own business.

**Comment by gregw** [comment]
@cowwoc client caches are always a lot more complex that you'd think. 
You can always use an established cache like squid as a proxy.
We also should consider looking at existing java HTTP caches and integrating them to our client rather than rolling our own

**Comment by cowwoc** [comment]
Squid is a great product, but it doesn't really jive with the idea of an embedded HTTP client. I'd like to release a Java client SDK for my server, and in that context asking users to use squid is probably a non-starter.

I wonder if Cloudflare's proxy would solve this kind of problem. It would be transparent to end-users and reduce unnecessary requests to the origin server. I'll probably give that a try at some point.

BTW, it looks like https://square.github.io/okhttp/ also includes client-side caching. Unfortunately, they're crawling towards full JPMS support :)

**Comment by gregw** [comment]
The other approach is to go with a really simple cache implementation that avoids most of the nuances of a full cache.    That would at least establish the space and then we could look at either implementing more cache functionality or embedding some other cache implementation.

**Comment by gregw** [comment]
I'm thinking that either I implement a simple cache in the next month or two.... or we say we are not going to implement this.

@cowwoc are you still interested.... what are your minimal requirements for a meaningful cache?  persistent?  OK to use a filesystem?

**Comment by cowwoc** [comment]
Sorry, I've buried my head in AI-related work for the past 9 months so it's hard for me to speak to this as a current requirement. The goal was/is to use `Cache-Control` heavily to improve performance. Using the file-system should be fine. I assume there should be a way for me to detect whether the cache hit/missed and retrieve the cached response body in case of a hit.
